File: Core/actor/actormanager.cpp
#include "actormanager.h"

#include <QDebug>

#include "actor/actormessage.h"
#include "value/client/clientvaluemanager.h"
#include "macros.h"

QLatin1String ActorManager::MANAGER_ID = QLatin1String("ActorManager");

ActorManager::ActorManager(QObject *parent) : ManagerBase(parent)
{
}

LogCat::LOGCAT ActorManager::logCat() {
    return LogCat::ACTOR;
}

void ActorManager::init(LocalConfig* config) {
    iDebug() << Q_FUNC_INFO;

    REQUIRE_MANAGER(CommunicationManagerBase);
    m_commManager = getManager<CommunicationManagerBase>(CommunicationManagerBase::MANAGER_ID);

    REQUIRE_MANAGER(ValueManagerBase);
    m_valueManager = getManager<ValueManagerBase>(ValueManagerBase::MANAGER_ID);
}

QString ActorManager::id() {
    return MANAGER_ID;
}

MessageBase::MESSAGE_TYPE ActorManager::getMessageType() {
    return MessageBase::MESSAGE_TYPE_ACTOR;
}

void ActorManager::handleReceivedMessage(MessageBase* msg) {
    ActorMessage* actorMessage = static_cast<ActorMessage*>(msg);
    iDebug() << Q_FUNC_INFO << actorMessage->fullId() << actorMessage->cmd();

    if (m_actors.contains(actorMessage->fullId())) {
        ActorBase* actor = m_actors.value(actorMessage->fullId());

        if (actorMessage->cmd() != actor::ACTOR_CMD_SET_VALUE) {        // value will be updated in actor explicitly
            if (actorMessage->rawValue().isValid()) {           // TODO: check if allowed, like  && actor->cmdSupported(actor::ACTOR_CMD_SET_VALUE)?
                iDebug() << "Updating value" << actor->fullId() << actorMessage->rawValue();
                actor->updateValue(actorMessage->rawValue(), false);
            }
        }

        actor->triggerCmd(actorMessage->cmd(), "Message received");

        /*
        if (!actor->isAsync()) {        // only publish if value is set synchronously
            iDebug() << "Publish value";
            getManager<ClientValueManager>(ClientValueManager::MANAGER_ID)->publishValue(actor);
        }*/
    } else {
        iWarning() << "Invalid actor" << actorMessage->fullId();
    }
}

void ActorManager::registerActor(ActorBase* actor, ValueManagerBase *valueManager) {
    qDebug() << Q_FUNC_INFO << actor->fullId();

    m_actors.insert(actor->fullId(), actor);
    /*
    Q_ASSERT(connect(actor, &ActorBase::cmdTriggered, this, [this, actor](actor::ACTOR_CMDS cmd) {
        // TODO
    }) != nullptr);
    */

    valueManager->registerValue(actor);
}

void ActorManager::publishCmd(ActorBase *actor, actor::ACTOR_CMDS cmd) {
    iDebug() << Q_FUNC_INFO << actor->fullId();

    ActorMessage msg(actor->valueGroup()->id(), actor->id(), QVariant(), cmd);
    m_commManager->sendMessage(msg);
}

void ActorManager::publishCmd(ActorBase *actor, actor::ACTOR_CMDS cmd, QVariant value) {
    iDebug() << Q_FUNC_INFO << actor->fullId();

    ActorMessage msg(actor->valueGroup()->id(), actor->id(), value, cmd);
    m_commManager->sendMessage(msg);
}

/*
void ActorManager::onCmdTriggered(ActorBase* actor, actor::ACTOR_CMDS cmd) {
    iDebug() << Q_FUNC_INFO << cmd;

    ActorMessage msg(actor->valueGroup()->id(), actor->id(), cmd);
    m_commManager->sendMessage(msg);
}
*/

ActorBase* ActorManager::getActor(QString fullId) {
    return m_actors.value(fullId);
}

ActorBase* ActorManager::getActor(QString valueGroupId, QString actorId) {
    return getActor(ValueBase::getFullId(valueGroupId, actorId));
}

ActorBase* ActorManager::getActor(ValueGroup *valueGroup, QString actorId) {
    return getActor(valueGroup->id(), actorId);
}

QList<ActorBase*> ActorManager::actors() {
    return m_actors.values();
}

File: AndroidApp/OshCore/src/main/java/com/osh/actor/ActorManager.java
package com.osh.actor;

import android.util.Log;

import java.util.HashMap;
import java.util.Map;

import com.osh.communication.MessageBase;
import com.osh.communication.MessageBase.MESSAGE_TYPE;
import com.osh.manager.IManagerRegistration;
import com.osh.manager.ManagerBase;
import com.osh.value.ValueManager;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ActorManager extends ManagerBase implements IActorManager {

	private static final Logger log = LoggerFactory.getLogger(ActorManager.class);

    protected Map<String, ActorBase> actors = new HashMap<>();

	public ActorManager(IManagerRegistration managerRegistration) {
		super("ActorManager", managerRegistration);
	}

	@Override
	public void registerActor(ActorBase actor) {
		actors.put(actor.getId(), actor);
	}

	@Override
	public MESSAGE_TYPE getMessageType() {
		return MESSAGE_TYPE.MESSAGE_TYPE_ACTOR;
	}

	@Override
	public void handleReceivedMessage(MessageBase msg) {
		if (msg instanceof ActorMessage) {
			if (actors.containsKey(((ActorMessage) msg).getFullId())) {
				ActorBase actor = actors.get(((ActorMessage) msg).getFullId());
				actor.triggerCmd(((ActorMessage) msg).getCmd(), "Message received");
			}
		} else {
			log.warn("Invalid actor: " + ((ActorMessage) msg).getFullId());
		}
	}

	@Override
	public void initComplete() {

	}

}
